File: gtci/16-subsets/python3/1-subsets.py
class Solution:
    def find_subsets(self, nums):
        """
        O(N * 2^N) time

        Every time we consider a number in the input array, we double the number of subsets
        by adding the number to all existing subsets.
        """
        subsets = [[]]

        for num in nums:
            n_subsets = len(subsets)
            for i in range(n_subsets):
                new_subset = subsets[i] + [num]
                subsets.append(new_subset)

        return subsets


if __name__ == "__main__":
    sol = Solution()

    assert sol.find_subsets([1, 3]) == [[], [1], [3], [1, 3]]
    assert sol.find_subsets([1, 5, 3]) == [
        [],
        [1],
        [5],
        [1, 5],
        [3],
        [1, 3],
        [5, 3],
        [1, 5, 3],
    ]

    print("All test cases passed.")

File: gtci/16-subsets/cpp/1-subsets.cpp
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  vector<vector<int>> findSubsets(const vector<int>& nums) {
    vector<vector<int>> subsets {{}};
    
    for (int num : nums) {
      const size_t nSubsets = subsets.size();
      for (size_t i = 0; i < nSubsets; i++) {
        vector<int> newSubset = subsets[i];
        newSubset.push_back(num);

        subsets.push_back(newSubset);
      }
    }

    return subsets;
  }
};    

int main() {
    Solution s;
    
    vector<vector<int>> result1 = s.findSubsets({1,2,3});
    assert(result1.size() == 8);
    assert(result1[0] == vector<int>({}));
    assert(result1[1] == vector<int>({1}));
    assert(result1[2] == vector<int>({2}));
    assert(result1[3] == vector<int>({1,2}));
    assert(result1[4] == vector<int>({3}));
    assert(result1[5] == vector<int>({1,3}));
    assert(result1[6] == vector<int>({2,3}));
    assert(result1[7] == vector<int>({1,2,3}));
    
    vector<vector<int>> result2 = s.findSubsets({0});
    assert(result2.size() == 2);
    assert(result2[0] == vector<int>({}));
    assert(result2[1] == vector<int>({0}));

    std::cout << "All test cases passed." << std::endl;
    
    return 0;
}
